File: M2FBX/M2FBX/Source/MTObject/MT_ObjectHandler.cpp
#include "MT_ObjectHandler.h"

#include "MT_Object.h"
#include "Common.h"

#include <cstdio>

template <class TType>
size_t Read(FILE* InStream, TType* Buffer)
{
	return fread(Buffer, sizeof(TType), 1, InStream);
}

template <class TType>
size_t Write(FILE* OutStream, TType Buffer)
{
	return fwrite(&Buffer, sizeof(TType), 1, OutStream);
}

MT_Object* MT_ObjectHandler::ReadObjectFromFile(const std::string& FileName)
{
	FILE* InStream = nullptr;
	fopen_s(&InStream, FileName.data(), "rb");

	if (!InStream)
	{
		// Failed
		return nullptr;
	}

	// Try Validate Header
	uint Magic = 0;
	fread(&Magic, sizeof(uint), 1, InStream);
	if (Magic != 55530573)
	{
		// TODO: Probably send to MT_Object to validate?
		// Invalid
		return nullptr;
	}

	// Begin reading the file
	MT_Object* NewObject = new MT_Object();
	NewObject->ObjectName = ReadString(InStream, NewObject->ObjectName);
	fread(&NewObject->ObjectFlags, sizeof(uint), 1, InStream);

	// Read LODs
	if (NewObject->HasObjectFlag(MT_ObjectFlags::HasLODs))
	{
		ReadLODFromFile(InStream, NewObject);
	}

	fclose(InStream);

	return NewObject;
}

void MT_ObjectHandler::WriteObjectToFile(const std::string& FileName, const MT_Object& Object)
{
	FILE* OutStream = nullptr;
	fopen_s(&OutStream, FileName.data(), "wb");

	if (!OutStream)
	{
		// Failed
		return;
	}

	// Write magic
	Write(OutStream, (uint)55530573);

	// Begin to Write
	WriteString(OutStream, Object.ObjectName);
	Write(OutStream, Object.ObjectFlags);

	// Write LODs
	if (Object.HasObjectFlag(MT_ObjectFlags::HasLODs))
	{
		WriteLODToFile(OutStream, Object);
	}
}

void MT_ObjectHandler::ReadLODFromFile(FILE* InStream, MT_Object* NewObject)
{
	// Read LODS Size
	uint NumLODs = 0;
	Read(InStream, &NumLODs);
	NewObject->LodObjects.resize(NumLODs);

	// Begin to read LOD
	MT_Lod* LodObject = new MT_Lod();
	Read(InStream, &LodObject->VertexDeclaration);

	// Read the Vertices
	uint NumVertices = 0;
	Read(InStream, &NumVertices);
	LodObject->Vertices.resize(NumVertices);

	for (uint i = 0; i < NumVertices; i++)
	{
		Vertex NewVertex = Vertex();
		if (LodObject->HasVertexFlag(VertexFlags::Position))
		{
			Read(InStream, &NewVertex.position);
		}
		if (LodObject->HasVertexFlag(VertexFlags::Normals))
		{
			Read(InStream, &NewVertex.normals);
		}
		if (LodObject->HasVertexFlag(VertexFlags::Tangent))
		{
			Read(InStream, &NewVertex.tangent);
		}
		if (LodObject->HasVertexFlag(VertexFlags::Skin))
		{
			Read(InStream, &NewVertex.boneIDs);
			Read(InStream, &NewVertex.boneWeights);
		}
		if (LodObject->HasVertexFlag(VertexFlags::Color))
		{
			Read(InStream, &NewVertex.color0);
		}
		if (LodObject->HasVertexFlag(VertexFlags::Color1))
		{
			Read(InStream, &NewVertex.color1);
		}
		if (LodObject->HasVertexFlag(VertexFlags::TexCoords0))
		{
			Read(InStream, &NewVertex.uv0);
		}
		if (LodObject->HasVertexFlag(VertexFlags::TexCoords1))
		{
			Read(InStream, &NewVertex.uv1);
		}
		if (LodObject->HasVertexFlag(VertexFlags::TexCoords2))
		{
			Read(InStream, &NewVertex.uv2);
		}
		if (LodObject->HasVertexFlag(VertexFlags::ShadowTexture))
		{
			Read(InStream, &NewVertex.uv3);
		}

		LodObject->Vertices[i] = NewVertex;
	}

	// Read the FaceGroups
	uint NumFaceGroups = 0;
	Read(InStream, &NumFaceGroups);
	LodObject->FaceGroups.resize(NumFaceGroups);
	for (uint i = 0; i < NumFaceGroups; i++)
	{
		// Read FaceGroup
		MT_FaceGroup FaceGroup = MT_FaceGroup();
		Read(InStream, &FaceGroup.StartIndex);
		Read(InStream, &FaceGroup.NumFaces);

		// Read MaterialInstance
		MT_MaterialInstance* MaterialInstance = new MT_MaterialInstance();
		Read(InStream, &MaterialInstance->MaterialFlags);
		MaterialInstance->Name = ReadString(InStream, MaterialInstance->Name);
		MaterialInstance->DiffuseTexture = ReadString(InStream, MaterialInstance->DiffuseTexture);
		FaceGroup.MaterialInstance = MaterialInstance;

		LodObject->FaceGroups[i] = FaceGroup;
	}

	// Read the Indices
	uint NumIndices = 0;
	Read(InStream, &NumIndices);
	for (uint i = 0; i < NumIndices/3; i++)
	{
		Int3 Triangle = {};
		Read(InStream, &Triangle);
		LodObject->Indices.push_back(Triangle);
	}

	NewObject->LodObjects[0] = *LodObject;
	fclose(InStream);
}

void MT_ObjectHandler::WriteLODToFile(FILE* OutStream, const MT_Object& Object)
{
	Write(OutStream, (uint)Object.LodObjects.size());

	for (int i = 0; i < Object.LodObjects.size(); i++)
	{
		const MT_Lod& LodInfo = Object.LodObjects[i];

		Write(OutStream, LodInfo.VertexDeclaration);
		Write(OutStream, (uint)LodInfo.Vertices.size());

		for (int x = 0; x < LodInfo.Vertices.size(); x++)
		{
			const Vertex& VertexInfo = LodInfo.Vertices[i];
			if (LodInfo.HasVertexFlag(VertexFlags::Position))
			{
				Write(OutStream, VertexInfo.position);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::Normals))
			{
				Write(OutStream, VertexInfo.normals);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::Tangent))
			{
				Write(OutStream, VertexInfo.tangent);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::Skin))
			{
				Write(OutStream, VertexInfo.boneIDs);
				Write(OutStream, VertexInfo.boneWeights);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::Color))
			{
				Write(OutStream, VertexInfo.color0);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::Color1))
			{
				Write(OutStream, VertexInfo.color1);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::TexCoords0))
			{
				Write(OutStream, VertexInfo.uv0);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::TexCoords1))
			{
				Write(OutStream, VertexInfo.uv1);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::TexCoords2))
			{
				Write(OutStream, VertexInfo.uv2);
			}
			if (LodInfo.HasVertexFlag(VertexFlags::ShadowTexture))
			{
				Write(OutStream, VertexInfo.uv3);
			}
		}

		// Write FaceGroups
		Write(OutStream, (uint)LodInfo.FaceGroups.size());
		for (uint i = 0; i < LodInfo.FaceGroups.size(); i++)
		{
			// Write the FaceGroup
			const MT_FaceGroup& FaceGroupInfo = LodInfo.FaceGroups[i];
			Write(OutStream, FaceGroupInfo.StartIndex);
			Write(OutStream, FaceGroupInfo.NumFaces);

			// Write the MaterialInstance
			const MT_MaterialInstance& MaterialInfo = *FaceGroupInfo.MaterialInstance;
			Write(OutStream, MaterialInfo.MaterialFlags);
			WriteString(OutStream, MaterialInfo.Name);
			WriteString(OutStream, MaterialInfo.DiffuseTexture);
		}

		// Multiply by 3 because indices are stored as Int3's
		Write(OutStream, (uint)LodInfo.Indices.size() * 3); 
		for (uint i = 0; i < (uint)LodInfo.Indices.size(); i++)
		{
			Write(OutStream, LodInfo.Indices[i]);
		}
	}
}

File: Mafia2Libs/ResourceTypes/ModelHelpers/ModelExporter/MT_ObjectHandler.cs
using System.IO;
using Utils.StringHelpers;
using Utils.Models;
using Utils.SharpDXExtensions;

namespace ResourceTypes.ModelHelpers.ModelExporter
{
    public static class MT_ObjectHandler
    {
        /** Begin Deserialize functions */
        public static MT_Object ReadObjectFromFile(string file)
        {
            using (BinaryReader reader = new BinaryReader(File.Open(file, FileMode.Open)))
            {
                MT_Object ModelObject = new MT_Object();

                // Read Header, make sure it is valid before continuing.
                string FileHeader = StringHelpers.ReadStringBuffer(reader, 3);
                byte FileVersion = reader.ReadByte();
                bool bIsHeaderValid = ModelObject.IsHeaderValid(FileHeader, FileVersion);
                if (!bIsHeaderValid)
                {
                    // Invalid header
                    return null;
                }

                // Read Meta-Data
                ModelObject.ObjectName = StringHelpers.ReadString8(reader);
                ModelObject.ObjectFlags = (MT_ObjectFlags)reader.ReadInt32();

                if (ModelObject.ObjectFlags.HasFlag(MT_ObjectFlags.HasLODs))
                {
                    uint NumLODs = reader.ReadUInt32();
                    ModelObject.Lods = new MT_Lod[NumLODs];
                    for (int i = 0; i < NumLODs; i++)
                    {
                        MT_Lod NewLOD = ReadLODFromFile(reader);
                        ModelObject.Lods[i] = NewLOD;
                    }
                }

                return ModelObject;
            }
        }

        private static MT_Lod ReadLODFromFile(BinaryReader reader)
        {
            MT_Lod NewObject = new MT_Lod();
            NewObject.VertexDeclaration = (VertexFlags)reader.ReadInt32();

            // Attempt to read Vertices
            uint NumVertices = reader.ReadUInt32();
            NewObject.Vertices = new Vertex[NumVertices];
            for(int i = 0; i < NumVertices; i++)
            {
                Vertex NewVertex = new Vertex();
                VertexFlags LocalDeclaration = NewObject.VertexDeclaration;

                // Read Vertex Information using LocalDeclaration
                #region ReadVertexInfo
                if (LocalDeclaration.HasFlag(VertexFlags.Position))
                {
                    NewVertex.Position = Vector3Extenders.ReadFromFile(reader);
                }

                if (LocalDeclaration.HasFlag(VertexFlags.Normals))
                {
                    NewVertex.Normal = Vector3Extenders.ReadFromFile(reader);
                }

                if (LocalDeclaration.HasFlag(VertexFlags.Tangent))
                {
                    NewVertex.Tangent = Vector3Extenders.ReadFromFile(reader);
                }

                if(LocalDeclaration.HasFlag(VertexFlags.Skin))
                {
                    NewVertex.BoneIDs = reader.ReadBytes(4);
                    NewVertex.BoneWeights = new float[4];
                    for(int z = 0; z < 4; z++)
                    {
                        NewVertex.BoneWeights[z] = reader.ReadSingle();
                    }
                }

                if(LocalDeclaration.HasFlag(VertexFlags.Color))
                {
                    NewVertex.Color0 = reader.ReadBytes(4);
                }

                if (LocalDeclaration.HasFlag(VertexFlags.Color1))
                {
                    NewVertex.Color1 = reader.ReadBytes(4);
                }

                if (LocalDeclaration.HasFlag(VertexFlags.TexCoords0))
                {
                    NewVertex.UVs[0] = Half2Extenders.ReadFromFile(reader);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.TexCoords1))
                {
                    NewVertex.UVs[1] = Half2Extenders.ReadFromFile(reader);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.TexCoords2))
                {
                    NewVertex.UVs[2] = Half2Extenders.ReadFromFile(reader);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.ShadowTexture))
                {
                    NewVertex.UVs[3] = Half2Extenders.ReadFromFile(reader);
                }
                #endregion

                NewObject.Vertices[i] = NewVertex;
            }

            // Read FaceGroups
            uint NumFaceGroups = reader.ReadUInt32();
            NewObject.FaceGroups = new MT_FaceGroup[NumFaceGroups];
            for(int i = 0; i < NumFaceGroups; i++)
            {
                // Attempt to read FaceGroup
                MT_FaceGroup NewFaceGroup = new MT_FaceGroup();
                NewFaceGroup.StartIndex = reader.ReadUInt32();
                NewFaceGroup.NumFaces = reader.ReadUInt32();

                // Read FaceGroup Material
                MT_MaterialInstance NewMaterial = new MT_MaterialInstance();
                NewMaterial.MaterialFlags = (MT_MaterialInstanceFlags)reader.ReadInt32();
                NewMaterial.Name = StringHelpers.ReadString8(reader);

                if (NewMaterial.MaterialFlags.HasFlag(MT_MaterialInstanceFlags.HasDiffuse))
                {
                    NewMaterial.DiffuseTexture = StringHelpers.ReadString8(reader);
                }

                NewFaceGroup.Material = NewMaterial;
                NewObject.FaceGroups[i] = NewFaceGroup;
            }

            // Read Indices
            uint NumIndices = reader.ReadUInt32();
            NewObject.Indices = new uint[NumIndices];
            for(int i = 0; i < NumIndices; i++)
            {
                NewObject.Indices[i] = reader.ReadUInt32();
            }

            return NewObject;
        }

        /** End Deserialize functions */
        /** Begin Serialize Functions */

        public static void WriteObjectToFile(BinaryWriter writer, MT_Object ModelObject)
        {
            // Write Generic Header
            StringHelpers.WriteString(writer, "MTO", false);
            writer.Write((byte)3);

            // Write Meta-Data
            StringHelpers.WriteString8(writer, ModelObject.ObjectName);
            writer.Write((int)ModelObject.ObjectFlags);

            if(ModelObject.ObjectFlags.HasFlag(MT_ObjectFlags.HasLODs))
            {
                // Write LODs
                writer.Write(ModelObject.Lods.Length);

                for(int i = 0; i < ModelObject.Lods.Length; i++)
                {
                    WriteLODToFile(writer, ModelObject.Lods[i]);
                }
            }
        }

        private static void WriteLODToFile(BinaryWriter writer, MT_Lod LodObject)
        {
            writer.Write((int)LodObject.VertexDeclaration);

            // Attempt to write vertices
            writer.Write(LodObject.Vertices.Length);
            for(int i = 0; i < LodObject.Vertices.Length; i++)
            {
                Vertex Vert = LodObject.Vertices[i];
                VertexFlags LocalDeclaration = LodObject.VertexDeclaration;

                #region WriteVertexInfo
                if (LocalDeclaration.HasFlag(VertexFlags.Position))
                {
                    Vert.Position.WriteToFile(writer);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.Normals))
                {
                    Vert.Normal.WriteToFile(writer);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.Tangent))
                {
                    Vert.Tangent.WriteToFile(writer);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.Skin))
                {
                    writer.Write(Vert.BoneIDs);
                    for (int z = 0; z < 4; z++)
                    {
                        writer.Write(Vert.BoneWeights[z]);
                    }
                }
                if (LocalDeclaration.HasFlag(VertexFlags.Color))
                {
                    writer.Write(Vert.Color0);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.Color1))
                {
                    writer.Write(Vert.Color1);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.TexCoords0))
                {
                    Vert.UVs[0].WriteToFile(writer);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.TexCoords1))
                {
                    Vert.UVs[1].WriteToFile(writer);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.TexCoords2))
                {
                    Vert.UVs[2].WriteToFile(writer);
                }
                if (LocalDeclaration.HasFlag(VertexFlags.ShadowTexture))
                {
                    Vert.UVs[3].WriteToFile(writer);
                }
                #endregion EndVertexInfo
            }

            writer.Write(LodObject.FaceGroups.Length);
            for(int i = 0; i < LodObject.FaceGroups.Length; i++)
            {
                // Write FaceGroup
                MT_FaceGroup FaceGroup = LodObject.FaceGroups[i];
                writer.Write(FaceGroup.StartIndex);
                writer.Write(FaceGroup.NumFaces);

                // Write Material Instance
                MT_MaterialInstance MaterialInstance = FaceGroup.Material;
                writer.Write((int)MaterialInstance.MaterialFlags);
                StringHelpers.WriteString8(writer, MaterialInstance.Name);

                if (MaterialInstance.MaterialFlags.HasFlag(MT_MaterialInstanceFlags.HasDiffuse))
                {
                    StringHelpers.WriteString8(writer, MaterialInstance.DiffuseTexture);
                }
            }

            writer.Write(LodObject.Indices.Length);
            for(int i = 0; i < LodObject.Indices.Length; i++)
            {
                writer.Write(LodObject.Indices[i]);
            }
        }
        /** End Serialize Functions */
    }
}
